### PR TITLE
Fix various frame-event misunderstandings

### PR DESCRIPTION
**Pull request by RAOF**
Some of our tests have been incorrect because of misunderstandings of the behaviour of various Wayland concepts (damage only applies to pending buffers, a frame callback gets fired at most once, and so on).

Fix these, and while we're at it add an explicit test for the Firefox-endlessly-spins-without-rendering-anything scenario we were kinda accidentally testing (because we were repeatedly sending frame requests without actually submitting anything).

Closes: #292

### CODE REVIEW

File: tests/frame_submission.cpp [hunk 1]
```diff
@@ -98,3 +99,48 @@ TEST_F(FrameSubmission, test_buffer_can_be_deleted_after_attached)
...
 }
+
+/* Firefox has a lovely habit of sending an endless stream of
+ * wl_surface.frame() requests (maybe it's trying to estimate vsync?!)
+ *
+ * If a compositor responds immediately to the frame on commit, Firefox
+ * ends up looping endlessly. If the compositor *doesn't* respond to the frame
+ * request, Firefox decides not to draw anything. 🤷‍♀️
+ */
+TEST_F(FrameSubmission, test_frame_callback_storm)
```
Comment (AlanGriffiths) [inline tests/frame_submission.cpp:110]: I don't think this test name tells us much:

    [  FAILED  ] VirtualPointerV1Test.test_frame_callback_storm

```suggestion
TEST_F(FrameSubmission, when_client_endlessly_requests_frame_then_callbacks_are_throttled)
```

File: tests/test_surface_events.cpp [hunk 2]
```diff
@@ -534,11 +534,18 @@ TEST_F(ClientSurfaceEventsTest, frame_timestamp_increases)
     usleep(10000);
 
     wl_surface_attach(surface, buffers[2], 0, 0);
+    surface.add_frame_callback(
+        [&](int time)
+        {
+            EXPECT_THAT(time, Gt(prev_frame_time));
+            prev_frame_time = time;
+            frame_callback_count++;
+        });
```
Comment (AlanGriffiths) [inline tests/test_surface_events.cpp:543]: Do we need a second copy of the same lambda?

File: tests/test_surface_events.cpp [hunk 3]
```diff
@@ -534,11 +534,18 @@ TEST_F(ClientSurfaceEventsTest, frame_timestamp_increases)
...
     wl_surface_commit(surface);
 
     client.dispatch_until([&frame_callback_count]()
         {
-            return frame_callback_count >= 2;
+            return frame_callback_count == 2;
```
Comment (AlanGriffiths) [inline tests/test_surface_events.cpp:548]: While we should have a check[*] that `frame_callback_count` is 2, this risks the test hanging if it is more.

[*] Checking this in a test called "frame_timestamp_increases" is wrong: it belongs in "when_two_frames_are_requested_the_two_callbacks_are_seen"

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Makes sens, but some nitpicking

File: tests/test_surface_events.cpp (hunk 2 above)
Comment (RAOF) [reply tests/test_surface_events.cpp]: No

File: tests/test_surface_events.cpp (hunk 3 above)
Comment (RAOF) [reply tests/test_surface_events.cpp:545]: I'm pretty sure it's impossible for `frame_callback_count` to exceed 2. We only call `add_frame_callback` twice, and each callback is one-shot. I'm not *against* making it `>=2`, but no compositor behaviour can cause it to exceed 2.

Comment (AlanGriffiths) [reply tests/test_surface_events.cpp:545]: :shrug:

**Review by AlanGriffiths — APPROVED** [review]

**Pull request merged**